jovo-platforms/jovo-platform-facebookmessenger/test/FBMessengerResponse.test.ts

```diff
@@ -0,0 +1,144 @@
+  });
+
+  afterAll(() => {
+    jest.clearAllMocks();
+  });
+
+  beforeEach(() => {
+    app = new App();
+
+    const messenger = new FacebookMessenger({
+      fetchUserProfile: false,
+      shouldOverrideAppHandle: true,
+    });
+
+    app.use(messenger);
+    t = messenger.makeTestSuite();
+  });
+
+  test('single request', async (done) => {
+    app.setHandler({
+      LAUNCH_TEST() {
+        // Single fb api call
+        this.$messengerBot?.tell('hello');
+      },
+    });
+    const req = await t.requestBuilder.launch();
+    app.handle(ExpressJS.dummyRequest(req));
+
+    app.on('response', (handleRequest: HandleRequest) => {
+      const messengerResponse = handleRequest.jovo!.$response;
+      expect(messengerResponse).toMatchObject({
+        message: [
+          {
+            message: { quick_replies: undefined, text: 'hello world' },
+            message_type: 'RESPONSE',
+            recipient: { id: '2321893584590078' },
+          },
+        ],
+      });
+      done();
+    });
+  });
+
+  test('multiple responses', async (done) => {
+    app.setHandler({
+      LAUNCH_TEST() {
+        // Two fb api calls
+        this.$messengerBot?.showText({
+          text: 'hello world 1',
+        });
+        this.$messengerBot?.showText({
+          text: 'hello world 2',
+        });
+      },
+    });
+
+    const request: JovoRequest = await t.requestBuilder.launch();
+    app.handle(ExpressJS.dummyRequest(request));
+
+    app.on('response', (handleRequest: HandleRequest) => {
+      const messengerResponses = handleRequest.jovo!.$response;
+      expect(messengerResponses).toMatchObject({
+        message: [
+          {
+            message: { quick_replies: undefined, text: 'hello world 1' },
+            message_type: 'RESPONSE',
+            recipient: { id: '2321893584590078' },
+          },
+          {
+            message: { quick_replies: undefined, text: 'hello world 2' },
+            message_type: 'RESPONSE',
+            recipient: { id: '2321893584590078' },
+          },
+        ],
+      });
+      done();
+    });
+  });
+
+  test('multiple events', async (done) => {
+    app.setHandler({
+      INTENT_1() {
+        this.$messengerBot?.showText({
+          text: 'hello world 1',
+        });
+      },
+      INTENT_2() {
+        this.$messengerBot?.showText({
+          text: 'hello world 2',
+        });
+      },
+    });
+
+    // Request will call INTENT_1 and INTENT_2 in one call
+    const request: JovoRequest = await t.requestBuilder.launch(multipleEventRequest);
+    app.handle(ExpressJS.dummyRequest(request));
+
+    // This hook should be called twice since there are two events are called
+    app.on('response', (handleRequest: HandleRequest) => {
+      const response = handleRequest.jovo!.$response;
+      const request: any = handleRequest.jovo?.$request;
+
+      const intentName = request.messaging[0].postback.payload;
+
+      if (intentName === 'INTENT_1') {
```

Comment:
Hey @m-ripper - Any suggestions on how to improve the testing for a request with multiple events? I couldn't find a better way...

Reply:
Shouldn't the response theoretically contain the messages of both events? In that case, matching the response against the expected response should be sufficient.

The expected response would contain both messages in this case.

Reply:
That would be
```
{
  "message": [
    {
      "message": { "quick_replies": undefined, "text": "hello world 1" },
      "message_type": "RESPONSE",
      "recipient": { "id": "2321893584590078" }
    },
    {
      "message": { "quick_replies": undefined, "text": "hello world 2" },
      "message_type": "RESPONSE",
      "recipient": { "id": "2321893584590078" }
    }
  ]
}
            message: { quick_replies: undefined, text: 'hello world 1' },
              message_type: 'RESPONSE',
              recipient: { id: '2321893584590078' },    
    },
    {
                  message: { quick_replies: undefined, text: 'hello world 2' },
              message_type: 'RESPONSE',
              recipient: { id: '2321893584590078' },
    }
  ]
}
```
with the current structure I think.

Reply:
Theoretically, you could dynamically check if the number in the text is the same as the number in the intent name via a regex. But I am not sure if this is cleaner from a testing perspective. Another option would be this:

```
    // Request will call INTENT_1 and INTENT_2 in one call
    const request: JovoRequest = await t.requestBuilder.launch(multipleEventRequest);
    const host = ExpressJS.dummyRequest(request);
    host.setResponse = function (response: any) {
      expect(response).toMatch({ message: [  ... ] }) // here the message-array should contain both messages
    }
    await app.handle(host);
```

Here, a test-Host is created and `setResponse` overwritten to match the response that is passed as the first argument. I haven't actually tested this, but I think this should work as well.

Reply:
I've tested the code above locally and found that `setResponse` gets called 3 times. The first 2 are the results of `INTENT_1` and `INTENT_2` and the last is strangely an empty array. The first two calls make sense, but the last call should theoretically contain the result of both intents in an array. However, the output of a running local server yields the correct result. Do you have any suggestions on how to get around the multiple calls during testing?

Reply:
Hey there, the reason why the response in `setResponse()` is empty, is due to overwriting the `setResponse()`-method set in `overrideAppHandle()`. The best solution to this is to set `setResponse()` in `overrideAppHandle()` after the merging took place:

```
          const hostCopy: Host = Object.create(host.constructor.prototype);
          for (const key in host) {
            if (host.hasOwnProperty(key)) {
              const value = (host as any)[key];
              if (key === '$request') {
                (hostCopy as any)[key] = entry;
              } else {
                (hostCopy as any)[key] =
                  typeof value === 'object' && value.constructor.name === 'Object'
                    ? { ...value }
                    : value;
              }
            }
          }
          // tslint:disable-next-line
          hostCopy.setResponse = async function (obj: any) {
            responses.push(obj);
          };
          promises.push(PROTOTYPE_BACKUP.call(this, hostCopy));
```
This will avoid similar errors in the future. 

The snippet above now looks like this and succeeded for me:
```
    const request: JovoRequest = await t.requestBuilder.launch(multipleEventRequest);
    const host = ExpressJS.dummyRequest(request);
    host.setResponse = async function (response: any) {
      expect(response).toEqual([
        {
          messages: [
            {
              message: { quick_replies: undefined, text: 'hello world 1' },
              message_type: 'RESPONSE',
              recipient: { id: '2321893584590078' },
            },
          ],
        },
        {
          messages: [
            {
              message: { quick_replies: undefined, text: 'hello world 2' },
              message_type: 'RESPONSE',
              recipient: { id: '2321893584590078' },
            },
          ],
        },
      ]); // here the message-array should contain both messages
      done();
    };
    await app.handle(host);
```

If you have questions, just let me know.

Reply:
Thanks again @m-ripper! That fixed the issue 😄